pkg/distributor/query.go

```diff
@@ -90,10 +90,14 @@ func (d *Distributor) GetIngesters(ctx context.Context) (ring.ReplicationSet, er
 	// If tenant uses shuffle sharding, we should only query ingesters which are
 	// part of the tenant's subring.
 	shardSize := d.limits.IngestionTenantShardSize(userID)
-	lookbackPeriod := d.cfg.ShuffleShardingLookbackPeriod
-
-	if shardSize > 0 && lookbackPeriod > 0 {
-		return d.ingestersRing.ShuffleShardWithLookback(userID, shardSize, lookbackPeriod, time.Now()).GetReplicationSetForOperation(ring.Read)
+	queryIngestersWithin := d.limits.QueryIngestersWithin(userID)
+
+	// Only enable shuffle sharding on the read path when `query-ingesters-within`
+	// is non-zero since otherwise we can't determine if an ingester should be part
+	// of a tenant's shuffle sharding subring (we compare its registration time with
+	// the lookback period).
+	if shardSize > 0 && d.cfg.ShuffleShardingIngestersEnabled && queryIngestersWithin > 0 {
```

Comment:
I don't think we need `ShuffleShardingIngestersEnabled` flag -- we know that shuffle sharding is enabled if `shardSize > 0`.

Previous use of `QueryIngestersWithin` to set `ShuffleShardingLookbackPeriod` was based on in-order samples. I think the situation is not so clear with OOO samples -- if OOO window is say 1 week, and we also set `QueryIngestersWithin` to one week, we don't necessarily want to set `ShuffleShardingLookbackPeriod` to the same time -- that would effectively disable shuffle sharding on read path in our setup (with weekly rollouts).

I don't have a good idea how to solve this. Perhaps @pracucci has?

Reply:
In my opinion, we should not use lookback period that's longer than block retention time on ingesters. So that can be our upper limit.

Reply:
This issue wasn't introduced by this change right? It's always been an issue with increasing query-ingesters-within to support longer OOO windows. 

Would OOO samples be queryable if shuffle sharding is enabled and the lookback period to the block retention time, since the block retention time is probably lower than the ooo time window?

Reply:
I agree with Peter that we need to address it. Let me wrap my mind around it and I will follow up with my feedback.

> that would effectively disable shuffle sharding on read path in our setup (with weekly rollouts).

Ingesters don't leave the ring during rollouts, so "with weekly rollouts" doesn't apply, but your point is valid and true anyway, and we need to address it.

Reply:
I agree with Peter. My take:
1. We still need `ShuffleShardingIngestersEnabled` because shuffle sharding can be selectively enabled on read and write path independently. The value of `ShuffleShardingIngestersEnabled` here should be the one from the querier's config.
2. Using the ingesters TSDB retention instead of "query ingesters within" is a good strategy as "lookback period" is a good strategy.

Reply:
Updated the shuffle sharding lookback logic and documentation in https://github.com/grafana/mimir/pull/4287/files/25b204a8e7b27c2900b12fe59159b6bcf4e097fe..0690e96be924614d88cf70e8371cece9b05423fc